File: docs/experiments/11-hpc-sweep-numpy-gap/hpc_sweep_numpy_gap.m
(* ==========================================================================
   Experiment 11 -- Fourth sweep: closing the distance to NumPy
   ==========================================================================

   Runs unmodified in Mathilda and in Mathematica 14. See README.md.

       Mathilda      ./Mathilda -file hpc_sweep_numpy_gap.m
       Mathematica   wolframscript -file hpc_sweep_numpy_gap.m
       Python        python3 hpc_sweep_numpy_gap.py

   WHAT IT MEASURES.  This sweep did NOT start from applications.  The third
   sweep had left two application kernels far behind NumPy and neither was a
   linear-algebra row, so the hypothesis was that the remaining cost was in
   PRIMITIVES -- and specifically in two categories no earlier sweep had
   probed: STRUCTURAL operations (an O(1) or memcpy operation walking a boxed
   list) and SCANS.

   So this file is a probe table, not an application: one operation per line,
   each against its NumPy equivalent, at 10^6 float64.

   THE CONTROL IS THE POINT.  The last two rows -- a plain product and a Log
   -- were already at the memory floor before this sweep, which is what says
   that nothing above them is about arithmetic.  Read the table from the
   bottom up: everything expensive is structural or a scan.
   ========================================================================== *)

SetAttributes[bench, HoldRest];

(* bench[label, expr] -- one untimed warm-up, then the MINIMUM of three timed
   runs.  The minimum, not the mean: we are measuring the cost of the work, and
   every source of noise on a loaded machine can only add. *)
bench[label_String, expr_] := Module[{ts},
  expr;
  ts = Table[First[AbsoluteTiming[expr]], {3}];
  Print[StringPadRight[label, 52], ToString[Round[1000. Min[ts], 0.001]], " ms"]
];

check[label_String, value_] :=
  Print[StringPadRight[label, 52], "check = ", value];


n = 10^6;
v = RandomReal[{0, 1}, n];
k5 = RandomReal[{0, 1}, 5];
im = RandomReal[{0, 1}, {1024, 1024}];
k55 = RandomReal[{0, 1}, {5, 5}];

Print["Experiment 11 -- structural operations, scans and convolution"];
Print[""];
Print["-- the structural family: pure element MOVES --"];

bench["First[v]   (an O(1) element read)", First[v]];
bench["Last[v]",                           Last[v]];
bench["Most[v]",                           Most[v]];
bench["Rest[v]",                           Rest[v]];
bench["Drop[v, 250]",                      Drop[v, 250]];
bench["Part[v, 250 ;; -250]",              v[[250 ;; -250]]];
bench["Reverse[v]",                        Reverse[v]];
bench["RotateLeft[v, 3]",                  RotateLeft[v, 3]];

Print[""];
Print["-- scans --"];

bench["Accumulate[v]",                     Accumulate[v]];
bench["Differences[v]",                    Differences[v]];
bench["FoldList[Max, First[v], Rest[v]]",  FoldList[Max, First[v], Rest[v]]];
bench["FoldList[Max[#1,#2] &, ...]",       FoldList[Max[#1, #2] &, First[v], Rest[v]]];
bench["FoldList[Plus, 0., v]",             FoldList[Plus, 0., v]];
bench["FoldList[0.98 #1 + 0.02 #2 &, ...] (EMA)",
      FoldList[0.98 #1 + 0.02 #2 &, 0., v]];

Print[""];
Print["-- convolution --"];

bench["ListConvolve[k5, v]",               ListConvolve[k5, v]];
bench["ListCorrelate[k55, im]  (1024^2)",  ListCorrelate[k55, im]];

Print[""];
Print["-- clipping --"];
bench["Clip[v, {0.2, 0.8}]",               Clip[v, {0.2, 0.8}]];

Print[""];
Print["-- THE CONTROL: already at the memory floor before this sweep --"];
bench["v v",                                v v];
bench["Log[v]",                             Log[v]];

File: docs/experiments/11-hpc-sweep-numpy-gap/hpc_sweep_numpy_gap.py
#!/usr/bin/env python3
"""Experiment 11 -- Fourth sweep: closing the distance to NumPy.

Runs the same probe table as ``hpc_sweep_numpy_gap.m``.  See ``README.md``.

    python3 hpc_sweep_numpy_gap.py

This file is the REFERENCE the sweep was measured against: one operation per
line, at 10^6 float64.  Several rows have a NumPy spelling that returns a VIEW
rather than a new array (``v[:-1]``, ``v[::-1]``); a view is O(1) and would
flatter NumPy against two systems that both materialise, so those rows force a
copy.  The comparison is of the same work, not of the same syntax.

THE CONTROL IS THE POINT.  The last two rows -- a plain product and a log --
were already at parity before the sweep, which is what says that nothing above
them is about arithmetic.
"""

import time

import numpy as np
from scipy import signal as spsig


def bench(label, fn, reps=3):
    """One untimed warm-up, then the MINIMUM of `reps` timed runs."""
    fn()
    ts = []
    for _ in range(reps):
        t0 = time.perf_counter()
        fn()
        ts.append(time.perf_counter() - t0)
    print("%-52s%s ms" % (label, round(1000.0 * min(ts), 3)))


def check(label, value):
    print("%-52scheck = %s" % (label, value))

N = 10 ** 6
v = np.random.rand(N)
k5 = np.random.rand(5)
im = np.random.rand(1024, 1024)
k55 = np.random.rand(5, 5)


def ema(a, alpha=0.02):
    """The general linear recurrence has no NumPy primitive; lfilter is the
    library answer and is what a NumPy programmer would reach for."""
    return spsig.lfilter([alpha], [1.0, -(1.0 - alpha)], a)


def main():
    print("Experiment 11 -- structural operations, scans and convolution")
    print("")
    print("-- the structural family: pure element MOVES --")
    bench("v[0]      (an O(1) element read)", lambda: v[0])
    bench("v[-1]", lambda: v[-1])
    bench("v[:-1].copy()", lambda: v[:-1].copy())
    bench("v[1:].copy()", lambda: v[1:].copy())
    bench("v[250:].copy()", lambda: v[250:].copy())
    bench("v[249:-249].copy()", lambda: v[249:-249].copy())
    bench("v[::-1].copy()", lambda: v[::-1].copy())
    bench("roll(v, -3)", lambda: np.roll(v, -3))

    print("")
    print("-- scans --")
    bench("cumsum(v)", lambda: np.cumsum(v))
    bench("diff(v)", lambda: np.diff(v))
    bench("maximum.accumulate(v)", lambda: np.maximum.accumulate(v))
    bench("add.accumulate(v)", lambda: np.add.accumulate(v))
    bench("lfilter (EMA)", lambda: ema(v))

    print("")
    print("-- convolution --")
    bench("convolve(v, k5)", lambda: np.convolve(v, k5, mode="valid"))
    bench("correlate2d(im, k55)  (1024^2)",
          lambda: spsig.correlate(im, k55, mode="valid"))

    print("")
    print("-- clipping --")
    bench("clip(v, 0.2, 0.8)", lambda: np.clip(v, 0.2, 0.8))

    print("")
    print("-- THE CONTROL: the memory floor --")
    bench("v * v", lambda: v * v)
    bench("log(v)", lambda: np.log(v))


if __name__ == "__main__":
    main()
